testonly/hammer/contents.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hammer
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian"
+)
+
+// mapContents is a complete copy of the map's contents at a particular revision.
+type mapContents struct {
+	data map[mapKey]string
+	rev  int64
+}
+type mapKey [sha256.Size]byte
+
+func (m *mapContents) empty() bool {
+	return len(m.data) == 0
+}
+
+// pickKey randomly select a key that already exists in a given copy of the
+// map's contents. Assumes that the selected copy is non-empty.
+func (m *mapContents) pickKey(prng rand.Source) []byte {
+	if m.empty() {
+		panic("internal error: can't pick a key, map data is empty!")
+	}
+
+	choice := intN(prng, len(m.data))
+	// Need sorted keys for reproduceability.
+	keys := make([]mapKey, 0)
+	for k := range m.data {
+		keys = append(keys, k)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return bytes.Compare(keys[i][:], keys[j][:]) == -1
+	})
+	for _, k := range keys {
```

Comment:
why not return `keys[choice][:]`? (with optional `if len(keys) < ...` - this slice is local so its length might well be cached)